Check bad syntax on dockerfile before building.

yongtang opened the pull request:
**\- What I did**

This fix tries to address the issue raised in #26453 where bad syntax on dockerfile is not checked before building, thus user has to wait before seeing error in dockerfile.

**\- How I did it**

This fix fixes the issue by evaluating all the instructions and check syntax before dockerfile is invoked actually.

**\- How to verify it**

All existing tests pass.

**\- Description for the changelog**

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes  #26453

Signed-off-by: Yong Tang yong.tang.github@outlook.com

yongtang committed:
Check bad syntax on dockerfile before building.

This fix tries to address the issue raised in 26453 where bad syntax
on dockerfile is not checked before building, thus user has to wait
before seeing error in dockerfile.

This fix fixes the issue by evaluating all the instructions and check
syntax before dockerfile is invoked actually.

All existing tests pass.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

cyphar commented on builder/dockerfile/evaluator.go:
```diff
@@ -201,3 +201,31 @@ func (b *Builder) dispatch(stepN int, stepTotal int, ast *parser.Node) error {
+// arg, env, etc., this syntax check will not be complete and could not replace
+// the runtime check. Instead, this function is only a helper that allows
+// user to find out the obvious error in Dockerfile earlier on.
+func (b *Builder) evaluateDispatch(ast *parser.Node) error {
```
IMO this should be called "checkDispatch", since you're not actually evaluating anything here.

yongtang replied:
Thanks @cyphar. The PR has been updated with name changed.

runcom commented:
@yongtang would be nice to have integration tests to check this behavior (and fix the old one probably)

duglin commented on builder/dockerfile/evaluator.go:
```diff
@@ -201,3 +201,31 @@ func (b *Builder) dispatch(stepN int, stepTotal int, ast *parser.Node) error {
+		return err
+	}
+
+	if cmd == "onbuild" {
```
While this is a trivial check I'd prefer if we didn't treat "ONBIULD" as special since this type of check could be done for pretty much all of the cmds, so we should be consistent and leave this function as just a check of the cmds themselves and nothing more - for now.

duglin replied:
ok, now I think I know why you did ONBUILD on its own... you wanted to check the next args to see if its a valid cmd.  So, why not do that here instead of in `processImageFrom()`?  I'd prefer if we did all of this simple syntax checking in one spot.

yongtang replied:
Thanks @duglin the PR has been updated with check done inside `checkDispatch()`

duglin commented on builder/dockerfile/internals.go:
```diff
@@ -435,14 +435,18 @@ func (b *Builder) processImageFrom(img builder.Image) error {
 				return fmt.Errorf("%s isn't allowed as an ONBUILD trigger", n.Value)
 			}
-
+			if err := b.evaluateDispatch(n); err != nil {
```
if I'm following, I'd prefer to undo this split here and just do the syntax check in the `evaluateDispatch()` func above.

duglin commented on cli/command/image/build.go:
```diff
@@ -293,6 +293,9 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
 
 	response, err := dockerCli.Client().ImageBuild(ctx, body, buildOptions)
 	if err != nil {
+		if options.quiet {
```
What's this change about?

yongtang replied:
The change is to adjust the case in `TestBuildNotVerboseFailure` because it requires stderr output in quiet mode and stdout+stderr output in verbose mode to be equal. Without the change the `Sending build context to Docker daemon 2.048 kB` will be missing from the output:

```
func (s *DockerSuite) TestBuildNotVerboseFailure(c *check.C) {
        // This test makes sure that -q works correctly when build fails by
        // comparing between the stderr output in quiet mode and in stdout
        // and stderr output in verbose mode
        tt := []struct {
                TestName  string
                BuildCmds string
        }{
                {"quiet_build_no_from_at_the_beginning", "RUN whoami"},
                {"quiet_build_unknown_instr", "FROMD busybox"},
        }

        for _, te := range tt {
                _, _, qstderr, qerr := buildImageWithStdoutStderr(te.TestName, te.BuildCmds, false, "-q", "--force-rm", "--rm")
                _, vstdout, vstderr, verr := buildImageWithStdoutStderr(te.TestName, te.BuildCmds, false, "--force-rm", "--rm")
                if verr == nil || qerr == nil {
                        c.Fatal(fmt.Errorf("Test [%s] expected to fail but didn't", te.TestName))
                }
                if qstderr != vstdout+vstderr {
                        c.Fatal(fmt.Errorf("Test[%s] expected that quiet stderr and verbose stdout are equal; quiet [%v], verbose [%v]", te.TestName, qstderr, vstdout+vstderr))
                }
    }
```

duglin commented:
Just a few questions/comments.  Can you also look to add some testcases?  Or do existing tests cover it?

yongtang commented:
Thanks @runcom @duglin for the review. The pull request has been updated and integration test cases have been added to the PR.

Please take a look and let me know if there are any issues.

duglin commented:
LGTM

vdemeester commented:
LGTM 🐸
Moving this to doc review as I think we should at least note something in the build docs (and in API changes too 👼)
/cc @thaJeztah

thaJeztah commented:
Yes, I think we could use some information that the builder performs a "basic" validation of the Dockerfile before starting the actual build. Perhaps;

- in the changelog (although strictly, it's not versioned, so we should probably mention that there); https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v125-api-changes
-  in the description of the endpoint https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.25.md#build-image-from-a-dockerfile (here as well; we may need to update old versions of the API docs as well, since it's not versioned)
- a mention of this in the "usage" section of the builder https://github.com/docker/docker/blob/master/docs/reference/builder.md#usage (thinking of adding it to the paragraph starting with "The Docker daemon runs the instructions ...", but open to suggestions)

yongtang committed:
Update documentation and change log to include the preliminary validation of dockerfile.

This commit updates documentation and change log to include
the preliminary validation of the dockerfile before instructions
in dockerfile is run one-by-one.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @thaJeztah. The docs and changelog have been updated. Please take a look and let me know if there are any issues.

thaJeztah commented on docs/reference/api/docker_remote_api_v1.18.md:
```diff
@@ -1201,6 +1201,11 @@ The archive may include any number of other files,
+Before the Docker daemon runs the instructions in the `Dockerfile`, it performs
+a preliminary validation of the `Dockerfile` and returns error if the syntax is
+incorrect. After that, each instruction is run one-by-one until the ID of the new
+image is output.
```
Was thinking;

```Markdown
The Docker daemon performs a preliminary validation of the `Dockerfile` before
starting the build, and returns an error if the syntax is incorrect. After that,
each instruction is run one-by-one until the ID of the new image is output.
```

Perhaps `preliminary` should be `cursory`?

@duglin @mstanleyjones could you have a look at my suggestions?

thaJeztah commented on docs/reference/api/docker_remote_api.md:
```diff
@@ -123,6 +123,7 @@ This section lists each version from latest to oldest.  Each listing includes a
 * `POST /containers/create/` and `POST /containers/(name)/update` now validates restart policies.
 * `POST /containers/create` now validates IPAMConfig in NetworkingConfig, and returns error for invalid IPv4 and IPv6 addresses (`--ip` and `--ip6` in `docker create/run`).
 * `POST /containers/create` now takes a `Mounts` field in `HostConfig` which replaces `Binds` and `Volumes`. *note*: `Binds` and `Volumes` are still available but are exclusive with `Mounts`
+* `POST /build` now performs a preliminary validation of the `Dockerfile` and returns error, before each instruction is run one-by-one. Note this is unversioned and applied to all previous remote API versions.
```
"one-by-one" is not needed here, we can simplify to

```Markdown
* `POST /build` now performs a preliminary validation of the `Dockerfile` before starting the build,
  and returns an error if the syntax is incorrect. Note that this change is _unversioned_ and applied
  to all API versions.
```

thaJeztah commented on docs/reference/builder.md:
```diff
@@ -68,6 +68,13 @@ add multiple `-t` parameters when you run the `build` command:
     $ docker build -t shykes/myapp:1.0.2 -t shykes/myapp:latest .
 
+Before the Docker daemon runs the instructions in the `Dockerfile`, it performs
+a preliminary validation of the `Dockerfile` and returns error if the syntax is incorrect:
```
s/error/an error/

thaJeztah reviewed (changes requested):
Thanks @yongtang, I left some comments (but open to suggestions from native english people :smile:)

thaJeztah commented:
Orthogonal to this PR, but @duglin what would you think about having;

- having the client perform a validate step first, before sending the build-context (validation could take place on the daemon; same endpoint with a query-parameter or a separate endpoint)
- perhaps a `docker build --validate-only` option?

I know https://github.com/docker/docker/issues/26453 is about validating the Dockerfile to prevent having to wait for the _build_ to complete, but in some scenarios, sending the build-context could actually take up a lot of time; validating the Dockerfile before sending could be a big timesaver for those situations.

duglin commented:
If we can do the fail-quick-syntax-check on the Dockerfile on the client, and do it easily, meaning it doesn't require a major redesign of the code then I think its a good idea.  My only concern is https://github.com/docker/docker/pull/12749.  I really think that INCLUDE needs to support env vars, and args, and I'm not sure we can do a simple syntax check on included files w/o processing the ENV and ARG cmds fully.  Of course https://github.com/docker/docker/pull/12749 could impact this PR too - just didn't get that far yet in my thinking about it....

yongtang commented:
Thanks @thaJeztah. The documentation has been updated. Please let me know if there are any issues.

thaJeztah commented:
docs LGTM, thanks!

vdemeester reviewed (approved):
docs LGTM 🐸

mcornella commented:
Thanks everybody!

thaJeztah commented:
Thanks @yongtang - let me know if you (with @duglin) want to work on the improvements suggested above (https://github.com/docker/docker/pull/26516#issuecomment-248868526). Perhaps we should open an issue for tracking that?

yongtang commented:
Thanks @duglin @thaJeztah. I opened a new issue #26877 to track the improvement proposal. I will see if I can open a PR soon.
